src/zkevm_specs/evm/table.py

```diff
@@ -0,0 +1,199 @@
+    # State circuit directly in their first row, and most of them will be
+    # read and checked to be agreed with caller before the beginning of call.
+    # It's not like transaction or bytecode that require specifically friendly
+    # layout for verification, so maintaining the consistency directly in
+    # RWTable seems more intuitive than creating another table for it.
+    RWCounterEndOfReversion = auto()  # to know at which point in the future we should revert
+    CallerCallId = auto()  # to know caller's id
+    TxId = auto()  # to know tx's id
+    Depth = auto()  # to know if call too deep
+    CallerAddress = auto()
+    CalleeAddress = auto()
+    CalldataOffset = auto()
+    CalldataLength = auto()
+    ReturndataOffset = auto()  # for callee to set returndata to caller's memeory
+    ReturndataLength = auto()
+    Value = auto()
+    Result = auto()  # to peek result in the future
+    IsPersistent = auto()  # to know if current call is within reverted call or not
+    IsStatic = auto()  # to know if state modification is within static call or not
+
+    # The following are used by caller to save its own CallState when it's
+    # going to dive into another call, and will be read out to restore caller's
+    # CallState in the end by callee.
+    # Note that stack and memory could also be included here, but since they
+    # need extra constraints on their data format, so we separate them to be
+    # different kinds of RWTableTag.
+    IsRoot = auto()
+    IsCreate = auto()
+    OpcodeSource = auto()
+    ProgramCounter = auto()
+    StackPointer = auto()
+    GasLeft = auto()
+    MemorySize = auto()
+    StateWriteCounter = auto()
+
+
+class Tables:
+    """
+    A collection of lookup tables used in EVM circuit.
+    """
+
+    fixed_table: Set[Tuple[
+        int,  # tag
+        int,  # value1
+        int,  # value2
+        int,  # value3
+    ]] = set(
+        [(FixedTableTag.Range32, i, 0, 0) for i in range(32)] +
+        [(FixedTableTag.Range64, i, 0, 0) for i in range(64)] +
+        [(FixedTableTag.Range256, i, 0, 0) for i in range(256)] +
+        [(FixedTableTag.Range512, i, 0, 0) for i in range(512)] +
+        [(FixedTableTag.Range1024, i, 0, 0) for i in range(1024)] +
+        [(FixedTableTag.InvalidOpcode, opcode, 0, 0) for opcode in invalid_opcodes()] +
+        [(FixedTableTag.StateWriteOpcode, opcode, 0, 0) for opcode in state_write_opcodes()] +
+        [(FixedTableTag.StackUnderflow, opcode, stack_pointer, 0) for (opcode, stack_pointer) in stack_underflow_pairs()] +
+        [(FixedTableTag.StackOverflow, opcode, stack_pointer, 0) for (opcode, stack_pointer) in stack_overflow_pairs()] +
+        [(FixedTableTag.OOGConstant, opcode, gas, 0) for (opcode, gas) in oog_constant_pairs()]
+    )
+    tx_table: Set[Tuple[
+        int,  # tx_id
+        int,  # tag
+        int,  # index (or 0)
+        int,  # value
+    ]]
+    bytecode_table: Set[Tuple[
```

Comment:
what do you think of add "is_code" to bytecode_table ,which indicate whether it is true code or not ? seems common requirement  for three scenarios regarding byte code table.

Reply:
Here is my consideration:
If we don't want to explicitly gather (copy) opcode into a single table from different sources, for opcode fetching, we might have 3 sources in different situation:
- When contract interaction, we lookup `bytecode_table` to read bytecode
- When contract creation in root call, we lookup `tx_table` to read transaction's calldata
- When contract creation in internal call, we lookup `rw_table` to read caller's memory

With an annotation `is_code` will be useful for handling opcode `JUMP*`, however, the source could other than `bytecode_table`, so only annotating it in `bytecode_table` seems not a general solution, we also need to annotate such information in `tx_table` and `rw_table`, and the latter one is more tricky since the memory would not always be layouted adjacently.

So currently I prefer to just look back 32 opcodes when handling `JUMP*`, what do you think?

Reply:
I think looking back 32 opcodes seems has problem,  the back 32th byte can be code or data as well, also need to check it.   thus looking back deeper requires,   do you think it is problem ?

Reply:
But `PUSH*` is up to `PUSH32`, why do we need to look back more than 32 opcodes?

Reply:
let's say see example:  push32  00 01 02 03 ... 1f
translate to byte code sequence is: 
7f  00 01. 02...  1f
the '7f' can also be data segment as it is maybe in previous push* section recursively.   right ?

Reply:
Ohh, you are right. Never thought about such false alarm case... Maybe we need to annotate `is_code` for them. Let me think about this and convert it into an issue for further discussion.

Could we add `is_code` in `bytecode_table` in further PR for more discussion?

Reply:
sure !